File: packages/swift/Sources/Anyagent/Anyagent.swift
// The anyagent binary as a Swift API: spawn `anyagent serve`, write command
// lines, route reply and event lines. Every rule lives in the binary; this
// file is a pipe (ticket 13, W1–W10).

import Foundation

// ---------------------------------------------------------------------------
// PUBLIC TYPES
// ---------------------------------------------------------------------------

/// What `open` and `generate` accept besides the agent: the `open` command's fields.
public struct OpenOptions: Encodable, Sendable {
    public var dir: String
    public var resume: String?
    public var fork: String?
    public var forkAt: String?
    public var permissionMode: PermissionMode?
    public var mcpServers: [McpServer]?
    public var configure: [String: ConfigValue]?

    public init(
        dir: String, resume: String? = nil, fork: String? = nil, forkAt: String? = nil,
        permissionMode: PermissionMode? = nil, mcpServers: [McpServer]? = nil, configure: [String: ConfigValue]? = nil
    ) {
        self.dir = dir
        self.resume = resume
        self.fork = fork
        self.forkAt = forkAt
        self.permissionMode = permissionMode
        self.mcpServers = mcpServers
        self.configure = configure
    }

    enum CodingKeys: String, CodingKey {
        case dir, resume, fork, configure
        case forkAt = "fork_at"
        case permissionMode = "permission_mode"
        case mcpServers = "mcp_servers"
    }
}

/// `kind`, `message`, and every extra field from the wire in `data` (W8).
public struct AnyagentError: Error, Sendable, CustomStringConvertible {
    public let kind: String
    public let message: String
    public let data: [String: JSONValue]

    public var description: String { "\(kind): \(message)" }

    init(kind: String, message: String, data: [String: JSONValue] = [:]) {
        self.kind = kind
        self.message = message
        self.data = data
    }

    /// From a wire `error` object.
    init(body: [String: JSONValue]) {
        var data = body
        kind = if case .string(let s) = data.removeValue(forKey: "kind") { s } else { "" }
        message = if case .string(let s) = data.removeValue(forKey: "message") { s } else { "" }
        self.data = data
    }
}

// ---------------------------------------------------------------------------
// RUNTIME: one `anyagent serve` process
// ---------------------------------------------------------------------------

/// The wire protocol this package speaks; the binary's hello must match.
let PROTOCOL: UInt32 = 1

/// One line from the binary, untagged: the fields present say what it is.
private struct Line: Decodable {
    var hello: Hello?
    var id: UInt64?
    var ok: JSONValue?
    var error: [String: JSONValue]?
    var event: Event?
    var session: String?
    var closed: String?
}

private struct Hello: Decodable {
    var `protocol`: UInt32
    var anyagent: String
}

/// One `anyagent serve` process. Build it with `Runtime.start`.
public actor Runtime {
    private let process = Process()
    private let stdin = Pipe()
    private var next: UInt64 = 1
    private var pending: [UInt64: Pending] = [:]
    private var sessions: [String: Session] = [:]
    private var dead: AnyagentError?
    private var hello: CheckedContinuation<Void, Error>?
    private var exited: Task<Int32, Never>?

    private struct Pending {
        let cont: CheckedContinuation<Reply, Error>
        let opens: Bool
    }

    private enum Reply {
        case ok(JSONValue)
        case session(Session)
    }

    /// Spawns the binary; returns after its hello line.
    /// `bin`: path to the binary; default `ANYAGENT_BIN`, then the app bundle's, then PATH.
    /// `mock`: a mock script (`packages/mock-scripts/*.json`): no real agents.
    /// `env`: environment for the binary and the agents it spawns; default this process's.
    public static func start(bin: String? = nil, mock: String? = nil, env: [String: String]? = nil) async throws -> Runtime {
        let rt = Runtime()
        try await rt.launch(bin: bin, mock: mock, env: env)
        return rt
    }

    public func discover() async throws -> DiscoveryReport {
        try await call(["cmd": "discover"])
    }

    public func probe(_ agent: AgentRef) async throws -> AgentDetails {
        try await call(["cmd": "probe", "agent": json(agent)])
    }

    public func planUsage(_ agent: AgentRef) async throws -> PlanUsage {
        try await call(["cmd": "plan_usage", "agent": json(agent)])
    }

    /// One-shot text with no session to manage: titles, commit messages.
    public func generate(_ agent: AgentRef, _ opts: OpenOptions, _ prompt: String) async throws -> String {
        try await call(fields(of: opts).merging(["cmd": "generate", "agent": json(agent), "prompt": .string(prompt)]) { $1 })
    }

    /// Opens a session. The Session is registered inside onLine (W1).
    public func open(_ agent: AgentRef, _ opts: OpenOptions) async throws -> Session {
        guard case .session(let session) = try await send(fields(of: opts).merging(["cmd": "open", "agent": json(agent)]) { $1 }, opens: true) else {
            throw AnyagentError(kind: "ProtocolFailed", message: "open replied without a session")
        }
        return session
    }

    /// Graceful and idempotent (W5): close stdin, wait up to 5 s, then kill.
    @discardableResult
    public func close() async -> Int32 {
        guard let exited else { return -1 }
        if dead == nil {
            try? stdin.fileHandleForWriting.close()
            let deadline = Task {
                try? await Task.sleep(for: .seconds(5))
                if !Task.isCancelled { killProcess() }
            }
            let code = await exited.value
            deadline.cancel()
            return code
        }
        return await exited.value
    }

    // Used by Session; not part of the API.
    /// Writes one command line; returns the reply's `ok` as T.
    func call<T: Decodable & Sendable>(_ fields: [String: JSONValue]) async throws -> T {
        guard case .ok(let ok) = try await send(fields) else {
            throw AnyagentError(kind: "ProtocolFailed", message: "unexpected reply shape")
        }
        return try decode(T.self, from: ok)
    }

    /// One raw line on stdin (tests use it for a bad frame).
    func write(line: String) {
        try? stdin.fileHandleForWriting.write(contentsOf: Data((line + "\n").utf8))
    }

    var pid: Int32 { process.processIdentifier }

    /// Writes one command line; the continuation gets the reply's `ok`, or the Session it opened.
    private func send(_ fields: [String: JSONValue], opens: Bool = false) async throws -> Reply {
        if let dead { throw dead }  // W4
        let id = next
        next += 1
        var frame = fields
        frame["id"] = .number(Double(id))
        let data = try JSONEncoder().encode(frame)
        return try await withCheckedThrowingContinuation { cont in
            pending[id] = Pending(cont: cont, opens: opens)
            write(line: String(decoding: data, as: UTF8.self))
        }
    }

    /// Spawns the binary, starts the reader, waits for the hello.
    private func launch(bin: String?, mock: String?, env: [String: String]?) async throws {
        signal(SIGPIPE, SIG_IGN)  // a write after the binary died throws EPIPE instead of ending the app
        let stdout = Pipe()
        process.executableURL = URL(fileURLWithPath: try resolveBinary(bin))
        process.arguments = mock.map { ["serve", "--mock", $0] } ?? ["serve"]
        if let env { process.environment = env }
        process.standardInput = stdin
        process.standardOutput = stdout
        let (status, report) = AsyncStream<Int32>.makeStream()
        process.terminationHandler = { report.yield($0.terminationStatus); report.finish() }
        try process.run()
        exited = Task {
            for await line in Runtime.lines(of: stdout.fileHandleForReading) { await onLine(line) }
            var code: Int32 = -1
            for await s in status { code = s }
            await onExit(code)
            return code
        }
        try await withCheckedThrowingContinuation { hello = $0 }
    }

    /// Routes one stdout line. Registers a session before the next line is read (W1).
    private func onLine(_ text: String) async {
        if dead != nil { return }
        guard let line = try? JSONDecoder().decode(Line.self, from: Data(text.utf8)) else {
            return abort("not a frame: \(text)")
        }
        if let hello = line.hello {
            if hello.protocol != PROTOCOL { return abort("protocol \(hello.protocol), this package speaks \(PROTOCOL)") }
            self.hello?.resume()
            self.hello = nil
        } else if let id = line.id, let p = pending.removeValue(forKey: id) {
            if let error = line.error {
                p.cont.resume(throwing: AnyagentError(body: error))
            } else if p.opens {
                p.cont.resume(with: Result { .session(try register(line.ok ?? .null)) })
            } else {
                p.cont.resume(returning: .ok(line.ok ?? .null))
            }
        } else if let event = line.event {
            await sessions[event.sessionId]?.push(event)
        } else if let id = line.session, let error = line.error {  // W3
            await sessions[id]?.fail(AnyagentError(body: error))
        } else if let id = line.closed {
            await sessions.removeValue(forKey: id)?.end()
        }
    }

    private func register(_ ok: JSONValue) throws -> Session {
        let info = try decode(SessionInfo.self, from: ok)
        let session = Session(self, info)
        sessions[info.id] = session
        return session
    }

    /// A binary that does not speak the protocol (W10): kill it; onExit fails the rest.
    private func abort(_ why: String) {
        dead = AnyagentError(kind: "ProtocolFailed", message: why)
        killProcess()
    }

    /// Process gone: fail everything still waiting (W4).
    private func onExit(_ code: Int32) async {
        let error = dead ?? AnyagentError(
            kind: "ProcessExited", message: "anyagent exited (\(code))", data: ["status": .string("\(code)"), "stderr": ""])
        dead = error
        hello?.resume(throwing: error)
        hello = nil
        for p in pending.values { p.cont.resume(throwing: error) }
        pending = [:]
        for session in sessions.values { await session.fail(error) }
        sessions = [:]
    }

    private func killProcess() {
        if process.isRunning { kill(process.processIdentifier, SIGKILL) }
    }

    /// stdout as lines: a blocking read loop on its own thread feeding a stream, in order.
    private static func lines(of handle: FileHandle) -> AsyncStream<String> {
        let fd = handle.fileDescriptor
        return AsyncStream { cont in
            Thread.detachNewThread {
                let handle = FileHandle(fileDescriptor: fd)
                var buffer = Data()
                while case let chunk = handle.availableData, !chunk.isEmpty {
                    buffer.append(chunk)
                    var start = buffer.startIndex
                    while let newline = buffer[start...].firstIndex(of: UInt8(ascii: "\n")) {
                        var line = buffer[start..<newline]
                        if line.last == UInt8(ascii: "\r") { line = line.dropLast() }
                        cont.yield(String(decoding: line, as: UTF8.self))
                        start = newline + 1
                    }
                    buffer.removeSubrange(buffer.startIndex..<start)
                }
                cont.finish()
            }
        }
    }
}

// ---------------------------------------------------------------------------
// SESSION: one open session
// ---------------------------------------------------------------------------

/// Unread events a session may hold before it is closed as lagging (W6).
let CAP = 4096

/// One open session: commands in, an ordered event stream out.
public actor Session {
    public nonisolated let id: String
    /// Live: replaced on every `SessionUpdated` (W2).
    public private(set) var info: SessionInfo
    /// Live: replaced on every `StatusChanged` (W2).
    public private(set) var status: SessionStatus
    private let rt: Runtime
    private var queue: [Event] = []
    private var waiters: [CheckedContinuation<Void, Never>] = []
    private var error: AnyagentError?
    private var done = false

    // Built by Runtime.open; not part of the API.
    init(_ rt: Runtime, _ info: SessionInfo) {
        self.rt = rt
        id = info.id
        self.info = info
        status = info.status ?? .idle
    }

    public func prompt(_ text: String, attachments: [String] = []) async throws -> Delivery {
        try await call(["cmd": "prompt", "text": .string(text), "attachments": .array(attachments.map { .string($0) })])
    }

    public func answer(_ request: String, _ answer: Answer) async throws {
        try await run(["cmd": "answer", "request": .string(request), "answer": json(answer)])
    }

    public func configure(_ option: String, _ value: ConfigValue) async throws {
        try await run(["cmd": "configure", "option": .string(option), "value": json(value)])
    }

    public func cancel(clearQueue: Bool = false) async throws {
        try await run(["cmd": "cancel", "clear_queue": .bool(clearQueue)])
    }

    public func dequeue(_ prompt: String) async throws {
        try await run(["cmd": "dequeue", "prompt": .string(prompt)])
    }

    public func rollback(turns: Int, scope: RollbackScope) async throws {
        try await run(["cmd": "rollback", "turns": .number(Double(turns)), "scope": json(scope)])
    }

    public func compact() async throws {
        try await run(["cmd": "compact"])
    }

    public func close() async throws {
        try await run(["cmd": "close"])
    }

    /// This session's events in order. Ends after `closed` or when the reading
    /// task is cancelled; throws once on a session error, then ends (W3).
    public nonisolated func events() -> AsyncThrowingStream<Event, Error> {
        AsyncThrowingStream { try await self.next() }
    }

    private func call<T: Decodable & Sendable>(_ fields: [String: JSONValue]) async throws -> T {
        var fields = fields
        fields["session"] = .string(id)
        return try await rt.call(fields)
    }

    private func run(_ fields: [String: JSONValue]) async throws {
        let _: JSONValue = try await call(fields)
    }

    /// The next event; nil at the end or once the reading task is cancelled; the error once, when the queue is drained.
    private func next() async throws -> Event? {
        while true {
            if !queue.isEmpty { return queue.removeFirst() }
            if let error {
                self.error = nil
                done = true
                throw error
            }
            if done || Task.isCancelled { return nil }
            await withTaskCancellationHandler {
                await withCheckedContinuation { waiters.append($0) }
            } onCancel: {
                Task { await self.wake() }
            }
        }
    }

    // push, fail and end are called by Runtime.onLine/onExit; not part of the API.
    /// Keeps info and status live (W2), applies the cap (W6).
    func push(_ ev: Event) {
        if error != nil || done { return }
        if case .sessionUpdated(let i) = ev.kind { info = i }
        if case .statusChanged(let s) = ev.kind { status = s }
        queue.append(ev)
        if queue.count > CAP {
            fail(AnyagentError(kind: "ConsumerLagged", message: "\(CAP) events unread"))
            Task { try? await close() }
            return
        }
        wake()
    }

    func fail(_ e: AnyagentError) {
        if error == nil, !done {
            error = e
            wake()
        }
    }

    func end() {
        done = true
        wake()
    }

    /// Resumes every waiting reader; each re-checks the queue, the error, and its own cancellation.
    private func wake() {
        for w in waiters { w.resume() }
        waiters = []
    }
}

// ---------------------------------------------------------------------------
// HELPERS
// ---------------------------------------------------------------------------

/// Any Encodable as a JSON value, for a command field.
func json<T: Encodable>(_ value: T) throws -> JSONValue {
    try JSONDecoder().decode(JSONValue.self, from: JSONEncoder().encode(value))
}

/// An Encodable's fields, for `open` and `generate`.
func fields<T: Encodable>(of value: T) throws -> [String: JSONValue] {
    if case .object(let o) = try json(value) { return o }
    return [:]
}

/// A reply's `ok` as the type the command returns.
func decode<T: Decodable & Sendable>(_ type: T.Type, from value: JSONValue) throws -> T {
    try JSONDecoder().decode(type, from: JSONEncoder().encode(value))
}

// ---------------------------------------------------------------------------
// INTERNAL: finding the binary
// ---------------------------------------------------------------------------

/// `bin`, then `ANYAGENT_BIN`, then the app bundle's auxiliary executable, then PATH.
func resolveBinary(_ bin: String?) throws -> String {
    let env = ProcessInfo.processInfo.environment
    if let explicit = bin ?? env["ANYAGENT_BIN"] { return explicit }
    #if os(macOS)
    if let bundled = Bundle.main.url(forAuxiliaryExecutable: "anyagent") { return bundled.path }
    #endif
    for dir in (env["PATH"] ?? "").split(separator: ":") where FileManager.default.isExecutableFile(atPath: "\(dir)/anyagent") {
        return "\(dir)/anyagent"
    }
    throw AnyagentError(
        kind: "NotInstalled",
        message: "no anyagent binary: bundle it in the app (Contents/MacOS/anyagent), set ANYAGENT_BIN, or put it on PATH",
        data: ["agent": "anyagent"])
}

File: packages/go/anyagent.go
// Package anyagent is the anyagent binary as a Go API: spawn `anyagent serve`,
// write command lines, route reply and event lines. Every rule lives in the
// binary; this file is a pipe (ticket 13, W1–W10).
package anyagent

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"iter"
	"os"
	"os/exec"
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// PUBLIC TYPES
// ---------------------------------------------------------------------------

// Options for Start.
type Options struct {
	Bin  string   // path to the binary; default ANYAGENT_BIN, then PATH
	Mock string   // a mock script (packages/mock-scripts/*.json): no real agents
	Env  []string // environment for the binary and the agents it spawns; default this process's
}

// OpenOptions is what Open and Generate accept besides the agent: the `open` command's fields.
type OpenOptions struct {
	Dir            string                 `json:"dir"`
	Resume         string                 `json:"resume,omitempty"`
	Fork           string                 `json:"fork,omitempty"`
	ForkAt         string                 `json:"fork_at,omitempty"`
	PermissionMode PermissionMode         `json:"permission_mode,omitempty"`
	McpServers     []McpServer            `json:"mcp_servers,omitempty"`
	Configure      map[string]ConfigValue `json:"configure,omitempty"`
}

// Error is `kind`, `message`, and every extra field from the wire in Data (W8).
type Error struct {
	Kind    string
	Message string
	Data    map[string]any
}

func (e *Error) Error() string { return e.Kind + ": " + e.Message }

func (e *Error) UnmarshalJSON(b []byte) error {
	var body map[string]any
	if err := json.Unmarshal(b, &body); err != nil {
		return err
	}
	e.Kind, _ = body["kind"].(string)
	e.Message, _ = body["message"].(string)
	delete(body, "kind")
	delete(body, "message")
	e.Data = body
	return nil
}

// ---------------------------------------------------------------------------
// RUNTIME: one `anyagent serve` process
// ---------------------------------------------------------------------------

// The wire protocol this package speaks; the binary's hello must match.
const protocol = 1

// Longest stdout line accepted; a tool result carrying a big diff is one line.
const lineLimit = 64 << 20

// One line from the binary, untagged: the fields present say what it is.
type line struct {
	Hello *struct {
		Protocol int `json:"protocol"`
	} `json:"hello"`
	ID      *uint64         `json:"id"`
	OK      json.RawMessage `json:"ok"`
	Error   *Error          `json:"error"`
	Event   *Event          `json:"event"`
	Session string          `json:"session"`
	Closed  string          `json:"closed"`
}

type pending struct {
	ch    chan reply
	opens bool
}

type reply struct {
	ok      json.RawMessage
	session *Session
	err     error
}

// Runtime is one `anyagent serve` process. Build it with Start.
type Runtime struct {
	cmd      *exec.Cmd
	stdin    io.WriteCloser
	wmu      sync.Mutex // one frame at a time on stdin
	mu       sync.Mutex // everything below
	next     uint64
	pending  map[uint64]pending
	sessions map[string]*Session
	dead     error
	helloed  bool
	hello    chan error
	exited   chan struct{} // closed by onExit, after code is set
	code     int
}

// Start spawns the binary; returns after its hello line.
func Start(opts Options) (*Runtime, error) {
	bin, err := resolveBinary(opts.Bin)
	if err != nil {
		return nil, err
	}
	args := []string{"serve"}
	if opts.Mock != "" {
		args = append(args, "--mock", opts.Mock)
	}
	rt := &Runtime{
		cmd: exec.Command(bin, args...), next: 1, pending: map[uint64]pending{}, sessions: map[string]*Session{},
		hello: make(chan error, 1), exited: make(chan struct{}),
	}
	rt.cmd.Env = opts.Env
	rt.cmd.Stderr = os.Stderr
	stdout, err := rt.cmd.StdoutPipe()
	if err != nil {
		return nil, err
	}
	if rt.stdin, err = rt.cmd.StdinPipe(); err != nil {
		return nil, err
	}
	if err := rt.cmd.Start(); err != nil {
		return nil, err
	}
	go rt.read(stdout) // sends the hello, or the exit error (W4)
	if err := <-rt.hello; err != nil {
		return nil, err
	}
	return rt, nil
}

func (rt *Runtime) Discover() (DiscoveryReport, error) {
	var report DiscoveryReport
	return report, rt.call(map[string]any{"cmd": "discover"}, &report)
}

// Probe: agent is a catalog id like "claude", or an AgentRef with Acp set.
func (rt *Runtime) Probe(agent any) (AgentDetails, error) {
	var details AgentDetails
	return details, rt.call(map[string]any{"cmd": "probe", "agent": agent}, &details)
}

func (rt *Runtime) PlanUsage(agent any) (PlanUsage, error) {
	var usage PlanUsage
	return usage, rt.call(map[string]any{"cmd": "plan_usage", "agent": agent}, &usage)
}

// Generate is one-shot text with no session to manage: titles, commit messages.
func (rt *Runtime) Generate(agent any, opts OpenOptions, prompt string) (string, error) {
	var text string
	cmd := fields(opts)
	cmd["cmd"], cmd["agent"], cmd["prompt"] = "generate", agent, prompt
	return text, rt.call(cmd, &text)
}

// Open opens a session. The Session is registered inside onLine (W1).
func (rt *Runtime) Open(agent any, opts OpenOptions) (*Session, error) {
	cmd := fields(opts)
	cmd["cmd"], cmd["agent"] = "open", agent
	r, err := rt.send(cmd, true)
	if err != nil {
		return nil, err
	}
	return r.session, nil
}

// Close is graceful and idempotent (W5): close stdin, wait up to 5 s, then kill. Returns the exit code.
func (rt *Runtime) Close() int {
	rt.mu.Lock()
	dead := rt.dead
	rt.mu.Unlock()
	if dead == nil {
		rt.stdin.Close()
		select {
		case <-rt.exited:
		case <-time.After(5 * time.Second):
			rt.cmd.Process.Kill()
		}
	}
	<-rt.exited
	return rt.code
}

// Writes one command line; the reply's `ok` is unmarshaled into out (nil to ignore it).
func (rt *Runtime) call(cmd map[string]any, out any) error {
	r, err := rt.send(cmd, false)
	if err != nil || out == nil || len(r.ok) == 0 || string(r.ok) == "null" {
		return err
	}
	return json.Unmarshal(r.ok, out)
}

// Registers the pending reply, then writes the frame. W4: a dead runtime fails at once.
func (rt *Runtime) send(cmd map[string]any, opens bool) (reply, error) {
	rt.mu.Lock()
	if rt.dead != nil {
		rt.mu.Unlock()
		return reply{}, rt.dead
	}
	id := rt.next
	rt.next++
	cmd["id"] = id
	ch := make(chan reply, 1)
	rt.pending[id] = pending{ch, opens}
	rt.mu.Unlock()
	data, err := json.Marshal(cmd)
	if err != nil {
		return reply{}, err
	}
	rt.write(data)
	r := <-ch
	return r, r.err
}

// One line on stdin. A write after death fails; onExit reports it.
func (rt *Runtime) write(line []byte) {
	rt.wmu.Lock()
	defer rt.wmu.Unlock()
	rt.stdin.Write(append(line, '\n'))
}

// The reader goroutine: routes every stdout line, then reports the exit.
func (rt *Runtime) read(stdout io.Reader) {
	sc := bufio.NewScanner(stdout)
	sc.Buffer(make([]byte, 64<<10), lineLimit)
	for sc.Scan() {
		rt.onLine(append([]byte(nil), sc.Bytes()...)) // a copy: a reply's RawMessage outlives the scanner's buffer
	}
	if err := sc.Err(); err != nil {
		rt.mu.Lock()
		rt.abort(err.Error())
		rt.mu.Unlock()
	}
	code := 0
	if err := rt.cmd.Wait(); err != nil {
		code = -1
		if exit, ok := err.(*exec.ExitError); ok {
			code = exit.ExitCode()
		}
	}
	rt.onExit(code)
}

// Routes one stdout line. Registers a session before the next line is read (W1).
func (rt *Runtime) onLine(raw []byte) {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	if rt.dead != nil {
		return
	}
	var l line
	if err := json.Unmarshal(raw, &l); err != nil {
		rt.abort("not a frame: " + string(raw))
		return
	}
	switch {
	case l.Hello != nil:
		if l.Hello.Protocol != protocol {
			rt.abort(fmt.Sprintf("protocol %d, this package speaks %d", l.Hello.Protocol, protocol))
		} else if !rt.helloed {
			rt.helloed = true
			rt.hello <- nil
		}
	case l.ID != nil:
		p, ok := rt.pending[*l.ID]
		if !ok {
			return
		}
		delete(rt.pending, *l.ID)
		switch {
		case l.Error != nil:
			p.ch <- reply{err: l.Error}
		case p.opens:
			s, err := rt.register(l.OK)
			p.ch <- reply{session: s, err: err}
		default:
			p.ch <- reply{ok: l.OK}
		}
	case l.Event != nil:
		if s := rt.sessions[l.Event.SessionID]; s != nil {
			s.push(l.Event)
		}
	case l.Session != "" && l.Error != nil: // W3
		if s := rt.sessions[l.Session]; s != nil {
			s.fail(l.Error)
		}
	case l.Closed != "":
		if s := rt.sessions[l.Closed]; s != nil {
			delete(rt.sessions, l.Closed)
			s.end()
		}
	}
}

// The Session for an `open` reply, registered before the reply is delivered.
func (rt *Runtime) register(ok json.RawMessage) (*Session, error) {
	var info SessionInfo
	if err := json.Unmarshal(ok, &info); err != nil {
		return nil, err
	}
	s := &Session{ID: info.ID, rt: rt, info: info, status: SessionStatusIdle, queue: make(chan *Event, queueCap)}
	if info.Status != nil {
		s.status = *info.Status
	}
	rt.sessions[info.ID] = s
	return s, nil
}

// A binary that does not speak the protocol (W10): kill it; onExit fails the rest. Called with mu held.
func (rt *Runtime) abort(why string) {
	rt.dead = &Error{Kind: "ProtocolFailed", Message: why}
	rt.cmd.Process.Kill()
}

// Process gone: fail everything still waiting (W4).
func (rt *Runtime) onExit(code int) {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	if rt.dead == nil {
		rt.dead = &Error{Kind: "ProcessExited", Message: fmt.Sprintf("anyagent exited (%d)", code), Data: map[string]any{"status": fmt.Sprint(code), "stderr": ""}}
	}
	if !rt.helloed {
		rt.helloed = true
		rt.hello <- rt.dead
	}
	for id, p := range rt.pending {
		p.ch <- reply{err: rt.dead}
		delete(rt.pending, id)
	}
	for id, s := range rt.sessions {
		s.fail(rt.dead)
		delete(rt.sessions, id)
	}
	rt.code = code
	close(rt.exited)
}

// ---------------------------------------------------------------------------
// SESSION: one open session
// ---------------------------------------------------------------------------

// Unread events a session may hold before it is closed as lagging (W6).
const queueCap = 4096

// Session is one open session: commands in, an ordered event stream out.
type Session struct {
	ID     string
	rt     *Runtime
	mu     sync.Mutex
	info   SessionInfo
	status SessionStatus
	queue  chan *Event // events; closed after the error (if any) is set
	err    error
	ended  bool
}

// Info is live: replaced on every SessionUpdated (W2).
func (s *Session) Info() SessionInfo {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.info
}

// Status is live: replaced on every StatusChanged (W2).
func (s *Session) Status() SessionStatus {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.status
}

func (s *Session) Prompt(text string, attachments ...string) (Delivery, error) {
	var delivery Delivery
	if attachments == nil {
		attachments = []string{}
	}
	return delivery, s.call(map[string]any{"cmd": "prompt", "text": text, "attachments": attachments}, &delivery)
}

func (s *Session) Answer(request string, answer Answer) error {
	return s.call(map[string]any{"cmd": "answer", "request": request, "answer": answer}, nil)
}

// Configure: value is a string or a bool.
func (s *Session) Configure(option string, value any) error {
	return s.call(map[string]any{"cmd": "configure", "option": option, "value": value}, nil)
}

func (s *Session) Cancel(clearQueue bool) error {
	return s.call(map[string]any{"cmd": "cancel", "clear_queue": clearQueue}, nil)
}

func (s *Session) Dequeue(prompt string) error {
	return s.call(map[string]any{"cmd": "dequeue", "prompt": prompt}, nil)
}

func (s *Session) Rollback(turns int, scope RollbackScope) error {
	return s.call(map[string]any{"cmd": "rollback", "turns": turns, "scope": scope}, nil)
}

func (s *Session) Compact() error {
	return s.call(map[string]any{"cmd": "compact"}, nil)
}

func (s *Session) Close() error {
	return s.call(map[string]any{"cmd": "close"}, nil)
}

// Events yields this session's events in order. Ends after `closed`; yields
// one error on a session error, then ends (W3). Ranging again continues where
// the last range stopped.
func (s *Session) Events() iter.Seq2[Event, error] {
	return func(yield func(Event, error) bool) {
		for ev := range s.queue {
			if !yield(*ev, nil) {
				return
			}
		}
		if err := s.takeErr(); err != nil {
			yield(Event{}, err)
		}
	}
}

func (s *Session) call(cmd map[string]any, out any) error {
	cmd["session"] = s.ID
	return s.rt.call(cmd, out)
}

// push, fail and end are called by Runtime.onLine/onExit; not part of the API.
// Keeps info and status live (W2), applies the cap (W6).
func (s *Session) push(ev *Event) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.ended {
		return
	}
	if ev.Kind.SessionUpdated != nil {
		s.info = *ev.Kind.SessionUpdated
	}
	if ev.Kind.StatusChanged != nil {
		s.status = *ev.Kind.StatusChanged
	}
	select {
	case s.queue <- ev:
	default:
		s.failLocked(&Error{Kind: "ConsumerLagged", Message: fmt.Sprintf("%d events unread", queueCap)})
		go s.Close()
	}
}

func (s *Session) fail(err error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.failLocked(err)
}

func (s *Session) failLocked(err error) {
	if !s.ended {
		s.ended, s.err = true, err
		close(s.queue)
	}
}

func (s *Session) end() {
	s.mu.Lock()
	defer s.mu.Unlock()
	if !s.ended {
		s.ended = true
		close(s.queue)
	}
}

func (s *Session) takeErr() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	err := s.err
	s.err = nil
	return err
}

// ---------------------------------------------------------------------------
// HELPERS
// ---------------------------------------------------------------------------

// An OpenOptions' fields, as the top-level fields of `open` and `generate`.
func fields(opts OpenOptions) map[string]any {
	data, _ := json.Marshal(opts)
	var m map[string]any
	json.Unmarshal(data, &m)
	return m
}

// ---------------------------------------------------------------------------
// INTERNAL: finding the binary
// ---------------------------------------------------------------------------

// Options.Bin, then ANYAGENT_BIN, then PATH.
func resolveBinary(bin string) (string, error) {
	if bin == "" {
		bin = os.Getenv("ANYAGENT_BIN")
	}
	if bin != "" {
		return bin, nil
	}
	path, err := exec.LookPath("anyagent")
	if err != nil {
		return "", &Error{Kind: "NotInstalled", Message: "no anyagent binary on PATH: ship it next to your app, set ANYAGENT_BIN, or pass Options.Bin", Data: map[string]any{"agent": "anyagent"}}
	}
	return path, nil
}
